Add an "initial project setup" script

pieterdavid opened the pull request:
Mostly for convenience (and to prove to @alesaggio that it's not that hard to change CMSSW versions ;-) ).
This also removes the duplication between the jenkins scripts and README.md (which now simply links to those), which should make it a bit less work to update the recipes.

pieterdavid committed:
Add an "initial project setup" script

Mostly for convenience. This also removes the duplication between the
jenkins scripts and README.md (which now simply links to those), which
should make it a bit easier to update those.

pieterdavid committed:
add link to bootstrap script

pieterdavid committed:
relative links

pieterdavid commented:
Sorry for pushing more commits after creating the PR, each of one which triggered a jenkins build... I cancelled the first two (jenkins doesn't read README.md, so there shouldn't be any risk).

OlivierBondu commented on setup_project_with_framework.sh:
```diff
@@ -0,0 +1,47 @@
+startdir=$(pwd)
+
+cmssw_arch="slc6_amd64_gcc530" ## should be in sync with CMSSW.arch
+cmssw_version="CMSSW_8_0_30"   ## should be in sync with CMSSW.release
```
why not read it from `CMSSW.arch` and `CMSSW.release` files then ? This would again avoid having to do it twice ?
I see your warnings below but I am not sure to see the use case, is it for whenever we would want to test an unstable recipe before it is merged ?

pieterdavid replied:
That's easy to do, but I wanted to make the "initial setup" as light as possible, and now this is done before the Framework is cloned. But there is a solution to that: I can clone it to a temporary directory, then read them from there, do everything up to the bootstrap, and then move it into place. I will update this

pieterdavid replied:
Having this for unstable recipes would be nice, I'll see if it is easy to implement by taking a single argument

OlivierBondu commented on setup_project_with_framework.sh:
```diff
@@ -0,0 +1,47 @@
+## move Framework to the right place
+llbbdir="${CMSSW_BASE}/src/cp3_llbb"
+mkdir -p "${llbbdir}"
+mv "${fwkdir_tmp}" "${llbbdir}/"
```
I'm not sure I like moving repos after checking out... basically the only thing that prevents doing the clone directly in the right place is the `git cms-init` if I am not mistaken, maybe adding one in this file should not hurt calling the `bootstrap_jenkins.sh` after ?

pieterdavid replied:
it's more tricky: `git cms-init` needs the `${CMSSW_BASE}/src` directory to be empty, so as long as one is there in `bootstrap_jenkins.sh` you cannot have Framework in the right place until after that is called.

pieterdavid replied:
moving a git repository should be fine, see https://stackoverflow.com/questions/17568543/git-add-doesnt-work/17747571#17747571 for details (we don't even use submodules)

OlivierBondu replied:
ok, fine then :) thanks !

OlivierBondu commented on setup_project_with_framework.sh:
```diff
@@ -0,0 +1,47 @@
+mkdir -p "${llbbdir}"
+mv "${fwkdir_tmp}" "${llbbdir}/"
+## compile
+scram b -j4
```
if this is supposed to be on `ingrid-ui2`, using only 4 cores is a bit of underkill, we don't compile so often this would disturb all the users ?

pieterdavid replied:
I'm not sure you gain much with >4... what do you usually use?

pieterdavid replied:
-j7 now :)

OlivierBondu reviewed (changes requested):
Thanks ! looks fine, I have some questions because I am not sure to understand everything, but nothing critical :)

pieterdavid committed:
clone to tmp and read CMSSW.arch and CMSSW.release from there

pieterdavid committed:
merge with frameworkBootstrap.sh

from
https://gist.github.com/blinkseb/ef079e273a5b28ca95348f62c4f6f2c5#file-frameworkBootstrap-sh

pieterdavid commented:
I integrated the nice features from @blinkseb 's [frameworkBootstrap.sh](https://gist.github.com/blinkseb/ef079e273a5b28ca95348f62c4f6f2c5#file-frameworkBootstrap-sh).

The main differences with respect to that are:
- changed branch to a named option, and added `--pr`
- kept the "checkout to temporary then move" strategy, which makes sure that `CMSSW.version`, `CMSSW.arch` and `framework_bootstraps.sh` correspond to the version that is being checked out
- changed `exit` to `return` (otherwise it exits the calling shell if it was sourced).

pieterdavid committed:
typo :-(

OlivierBondu commented:
Thanks !
